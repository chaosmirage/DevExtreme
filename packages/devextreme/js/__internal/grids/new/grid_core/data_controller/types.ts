export type DataObject = Record<string, unknown>;
export type Key = unknown;
export type KeyExpr = unknown;
