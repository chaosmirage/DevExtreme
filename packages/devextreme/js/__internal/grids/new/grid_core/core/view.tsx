/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable @typescript-eslint/no-this-alias */
/* eslint-disable @typescript-eslint/no-use-before-define */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable max-classes-per-file */
/* eslint-disable spellcheck/spell-checker */
import type { Subscription, SubsGets } from '@ts/core/reactive/index';
import { toSubscribable } from '@ts/core/reactive/index';
import { Component, type ComponentType, render } from 'inferno';
import { hydrate } from 'inferno-hydrate';
// import { renderToString } from 'inferno-server';

export abstract class View<T extends {}> {
  private inferno: undefined | ComponentType;

  private props?: T;

  private firstRender = true;

  protected abstract component: ComponentType<T>;

  protected abstract getProps(): SubsGets<T>;

  public render(root: Element): Subscription {
    const ViewComponent = this.component;
    return toSubscribable(this.getProps()).subscribe((props: T) => {
      this.props = props;
      if (this.firstRender) {
        this.firstRender = false;
        // @ts-expect-error
        hydrate(<ViewComponent {...props}/>, root);
      } else {
        // @ts-expect-error
        render(<ViewComponent {...props}/>, root);
      }
    });
  }

  public asInferno(): ComponentType {
    // eslint-disable-next-line no-return-assign
    return this.inferno ??= this._asInferno();
  }

  private _asInferno() {
    const view = this;

    interface State {
      props: T;
    }

    return class InfernoView extends Component<{}, State> {
      private subscription?: Subscription;

      private readonly viewProps: SubsGets<T>;

      constructor() {
        super();
        this.viewProps = view.getProps();
        this.state = {
          props: view.getProps().unreactive_get(),
        };
      }

      public render(): JSX.Element | undefined {
        const ViewComponent = view.component;
        // @ts-expect-error
        return <ViewComponent {...this.state!.props}/>;
      }

      public componentDidMount(): void {
        this.subscription = toSubscribable(this.viewProps).subscribe((props) => {
          this.state ??= {
            props,
          };

          if (this.state.props !== props) {
            this.setState({ props });
          }
        });
      }

      public componentWillUnmount(): void {
        this.subscription!.unsubscribe();
      }
    };
  }
}
