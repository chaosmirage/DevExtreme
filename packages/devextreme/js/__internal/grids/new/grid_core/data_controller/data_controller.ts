/* eslint-disable @typescript-eslint/no-invalid-void-type */
/* eslint-disable no-param-reassign */
/* eslint-disable spellcheck/spell-checker */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import type { DataSource } from '@js/common/data';
import ArrayStore from '@js/common/data/array_store';
import type { SubsGets } from '@ts/core/reactive/index';
import {
  computed, effect, state,
} from '@ts/core/reactive/index';
import { createPromise } from '@ts/core/utils/promise';

// import { EditingController } from '../editing/controller';
// import type { Change } from '../editing/types';
import { OptionsController } from '../options_controller/options_controller';
import type { DataObject, Key } from './types';
import {
  getLocalLoadOptions,
  getStoreLoadOptions,
  isCustomStore,
  isLocalStore,
  normalizeDataSource,
  normalizeLocalOptions,
  normalizeRemoteOptions,
  updateItemsImmutable,
} from './utils';

export class DataController {
  private readonly pendingLocalOperations = {};

  private readonly loadedPromise = createPromise<void>();

  private readonly dataSourceConfiguration = this.options.oneWay('dataSource');

  private readonly keyExpr = this.options.oneWay('keyExpr');

  public readonly dataSource = computed(
    (dataSourceLike, keyExpr) => normalizeDataSource(dataSourceLike, keyExpr),
    [this.dataSourceConfiguration, this.keyExpr],
  );

  // TODO
  private readonly cacheEnabled = this.options.oneWay('cacheEnabled');

  private readonly pagingEnabled = this.options.twoWay('paging.enabled');

  public readonly pageIndex = this.options.twoWay('paging.pageIndex');

  public readonly pageSize = this.options.twoWay('paging.pageSize');

  private readonly remoteOperations = this.options.oneWay('remoteOperations');

  private readonly onDataErrorOccurred = this.options.action('onDataErrorOccurred');

  private readonly _items = state<DataObject[]>([]);

  public readonly items: SubsGets<DataObject[]> = this._items;

  private readonly _totalCount = state(0);

  public readonly totalCount: SubsGets<number> = this._totalCount;

  public readonly isLoading = state(false);

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public readonly filter = this.options.twoWay('filterValue');

  // public itemsWithChanges = computed(
  //   (items, changes: Change[] | undefined) => items.map((item) => (changes ?? []).filter(
  //     (change) => change.key === this.getDataKey(item),
  //   ).reduce((p, v) => ({
  //     // @ts-expect-error
  //     ...item, ...v.data,
  //   }), item)),
  //   [this.items, this.editing.changes],
  // );

  public readonly pageCount = computed(
    (totalCount, pageSize) => Math.ceil(totalCount / pageSize),
    [this.totalCount, this.pageSize],
  );

  private readonly normalizedRemoteOptions = computed(
    (remoteOperations, dataSource) => {
      const store = dataSource.store();
      return normalizeRemoteOptions(remoteOperations, isLocalStore(store), isCustomStore(store));
    },
    [this.remoteOperations, this.dataSource],
  );

  private readonly normalizedLocalOperations = computed(
    (normalizedRemoteOperations) => normalizeLocalOptions(normalizedRemoteOperations),
    [this.normalizedRemoteOptions],
  );

  public static dependencies = [OptionsController] as const;

  constructor(
    private readonly options: OptionsController,
  ) {
    effect(
      (dataSource) => {
        const changedCallback = (e?): void => {
          this.onChanged(dataSource, e);
        };
        const loadingChangedCallback = (): void => {
          this.isLoading.update(dataSource.isLoading());
        };
        const loadErrorCallback = (error: string): void => {
          const callback = this.onDataErrorOccurred.unreactive_get();
          callback({ error });
          changedCallback();
        };
        const customizeStoreLoadOptionsCallback = (e): void => {
          const localOptions = this.normalizedLocalOperations.unreactive_get();
          this.pendingLocalOperations[e.operationId] = getLocalLoadOptions(
            e.storeLoadOptions,
            localOptions,
          );
          e.storeLoadOptions = getStoreLoadOptions(
            e.storeLoadOptions,
            localOptions,
          );
        };

        const dataLoadedCallback = (e): void => {
          /*
            We use Deffered here because the code below is synchronous.
            customizeLoadResult callback does not support async code.
          */
          new ArrayStore(e.data).load(this.pendingLocalOperations[e.operationId]).done((data) => {
            e.data = data;
          }).fail((error) => {
            // @ts-expect-error
            e.data = new Deferred().reject(error);
          });
          this.pendingLocalOperations[e.operationId] = undefined;
        };

        if (dataSource.isLoaded()) {
          changedCallback();
        }
        dataSource.on('changed', changedCallback);
        dataSource.on('loadingChanged', loadingChangedCallback);
        dataSource.on('loadError', loadErrorCallback);

        // @ts-expect-error
        dataSource.on('customizeStoreLoadOptions', customizeStoreLoadOptionsCallback);
        // @ts-expect-error
        dataSource.on('customizeLoadResult', dataLoadedCallback);

        return (): void => {
          dataSource.off('changed', changedCallback);
          dataSource.off('loadingChanged', loadingChangedCallback);
          dataSource.off('loadError', loadErrorCallback);

          // @ts-expect-error
          dataSource.off('customizeStoreLoadOptions', customizeStoreLoadOptionsCallback);
          // @ts-expect-error
          dataSource.off('customizeLoadResult', dataLoadedCallback);
        };
      },
      [this.dataSource],
    );

    effect(
      () => {
        if (this.dataSource.unreactive_get().isLoaded()) {
          // eslint-disable-next-line @typescript-eslint/no-floating-promises
          this.dataSource.unreactive_get().load();
        }
      },
      [this.normalizedRemoteOptions],
    );

    effect(
      (dataSource, pageIndex, pageSize, filter, pagingEnabled) => {
        let someParamChanged = false;
        if (dataSource.pageIndex() !== pageIndex) {
          dataSource.pageIndex(pageIndex);
          someParamChanged ||= true;
        }
        if (dataSource.pageSize() !== pageSize) {
          dataSource.pageSize(pageSize);
          someParamChanged ||= true;
        }
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-boolean-literal-compare
        if (dataSource.requireTotalCount() !== true) {
          dataSource.requireTotalCount(true);
          someParamChanged ||= true;
        }
        if (dataSource.filter() !== filter) {
          dataSource.filter(filter);
          someParamChanged ||= true;
        }
        if (dataSource.paginate() !== pagingEnabled) {
          dataSource.paginate(pagingEnabled);
          someParamChanged ||= true;
        }

        if (someParamChanged || !dataSource.isLoaded()) {
          // eslint-disable-next-line @typescript-eslint/no-floating-promises
          dataSource.load();
        }
      },
      [this.dataSource, this.pageIndex, this.pageSize, this.filter, this.pagingEnabled],
    );
  }

  private onChanged(dataSource: DataSource, e): void {
    let items = dataSource.items() as DataObject[];

    if (e?.changes) {
      items = this._items.unreactive_get();
      items = updateItemsImmutable(items, e.changes, dataSource.store());
    }

    this._items.update(items);
    this.pageIndex.update(dataSource.pageIndex());
    this.pageSize.update(dataSource.pageSize());
    this._totalCount.update(dataSource.totalCount());
    this.loadedPromise.resolve();
  }

  public getDataKey(data: DataObject): Key {
    return this.dataSource.unreactive_get().store().keyOf(data);
  }

  public waitLoaded(): Promise<void> {
    return this.loadedPromise.promise;
  }
}
