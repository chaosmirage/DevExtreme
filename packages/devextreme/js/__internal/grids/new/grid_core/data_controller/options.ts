import type { DataSourceLike } from '@js/data/data_source';

interface PagingOptions {
  enabled?: boolean;
  pageSize?: number;
  pageIndex?: number;
}

interface RemoteOperationsOptions {
  filtering?: boolean;
  paging?: boolean;
  sorting?: boolean;
  summary?: boolean;
}

export interface Options {
  cacheEnabled?: boolean;
  dataSource?: DataSourceLike<unknown>;
  dateSerializationFormat?: string;
  keyExpr?: string | string[];
  onDataErrorOccurred?: unknown;
  paging?: PagingOptions;
  remoteOperations?: RemoteOperationsOptions | boolean;
}

export const defaultOptions = {
  paging: {
    enabled: true,
    pageSize: 6,
    pageIndex: 0,
  },
  remoteOperations: {
    filtering: false,
    paging: false,
    sorting: false,
    summary: false,
  },
} satisfies Options;
