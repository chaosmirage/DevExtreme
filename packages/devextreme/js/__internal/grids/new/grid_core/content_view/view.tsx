/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable spellcheck/spell-checker */
import type dxScrollable from '@js/ui/scroll_view/ui.scrollable';
import type { ScrollEventInfo } from '@js/ui/scroll_view/ui.scrollable';
import { combined, computed, state } from '@ts/core/reactive/index';
import { ColumnsController } from '@ts/grids/new/grid_core/columns_controller/columns_controller';
import type { Column } from '@ts/grids/new/grid_core/columns_controller/types';
import { View } from '@ts/grids/new/grid_core/core/view';
import { DataController } from '@ts/grids/new/grid_core/data_controller/index';
import { ErrorController } from '@ts/grids/new/grid_core/error_controller/error_controller';
import { createRef } from 'inferno';

import { OptionsController } from '../options_controller/options_controller';

export abstract class ContentView<TProps extends {}> extends View<TProps> {
  private readonly isNoData = computed(
    (isLoading, items) => !isLoading && items.length === 0,
    [this.dataController.isLoading, this.dataController.items],
  );

  public readonly items = computed(
    (dataItems, columns: Column[]) => dataItems.map(
      (item) => this.columnsController.createDataRow(
        item,
        columns,
      ),
    ),
    [this.dataController.items, this.columnsController.visibleColumns],
  );

  public readonly scrollableRef = createRef<dxScrollable>();

  public loadingText = this.options.twoWay('loadPanel.message');

  protected readonly viewportHeight = state(0);

  protected readonly scrollTop = state(0);

  protected readonly width = state(0);

  public static dependencies = [
    DataController, OptionsController, ErrorController, ColumnsController,
  ] as const;

  constructor(
    protected readonly dataController: DataController,
    protected readonly options: OptionsController,
    protected readonly errorController: ErrorController,
    protected readonly columnsController: ColumnsController,

  ) {
    super();
  }

  protected getBaseProps() {
    return {
      loadPanelProps: computed(
        (visible, loadPanel) => ({
          ...loadPanel,
          visible,
        }),
        [
          this.dataController.isLoading,
          this.options.oneWay('loadPanel'),
        ],
      ),
      noDataTextProps: combined({
        text: this.options.oneWay('noDataText'),
        template: this.options.template('noDataTemplate'),
        visible: this.isNoData,
      }),
      errorRowProps: combined({
        enabled: this.options.oneWay('errorRowEnabled'),
        errors: this.errorController.errors,
      }),
      onWidthChange: this.width.update.bind(this.width),
      virtualScrollingProps: combined({
        heightUp: 0,
        heightDown: 0,
        // heightUp: computed((virtualState) => virtualState.virtualTop, [this.virtualState]),
        // heightDown: computed((virtualState) => virtualState.virtualBottom, [this.virtualState]),
      }),
      onViewportHeightChange: this.viewportHeight.update.bind(this.viewportHeight),
      scrollableRef: this.scrollableRef,
      scrollableProps: combined({
        onScroll: this.onScroll.bind(this),
        direction: 'both' as const,
        scrollTop: this.scrollTop,
        scrollByContent: this.options.oneWay('scrolling.scrollByContent'),
        scrollByThumb: this.options.oneWay('scrolling.scrollByThumb'),
        showScrollbar: this.options.oneWay('scrolling.showScrollbar'),
        useNative: computed(
          (useNative) => (useNative === 'auto' ? undefined : useNative),
          [this.options.oneWay('scrolling.useNative')],
        ),
      }),
    };
  }

  private onScroll(e: ScrollEventInfo<unknown>): void {
    this.scrollTop.update(e.scrollOffset.top);
  }
}
