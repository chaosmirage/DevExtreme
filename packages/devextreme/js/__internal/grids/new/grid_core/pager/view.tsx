/* eslint-disable spellcheck/spell-checker */
import type { SubsGets } from '@ts/core/reactive/index';
import { combined, computed } from '@ts/core/reactive/index';

import { View } from '../core/view';
import { DataController } from '../data_controller/index';
import { OptionsController } from '../options_controller/options_controller';
import type { PagerProps } from './pager';
import { PagerView as Pager } from './pager';
import { calculatePageSizes, isVisible } from './utils';

export class PagerView extends View<PagerProps> {
  protected override component = Pager;

  public static dependencies = [DataController, OptionsController] as const;

  private readonly pageSizesConfig = this.options.oneWay('pager.allowedPageSizes');

  private readonly allowedPageSizes = computed(
    (pageSizesConfig, pageSize) => calculatePageSizes(
      this.allowedPageSizes?.unreactive_get(),
      pageSizesConfig,
      pageSize,
    ),
    [this.pageSizesConfig, this.dataController.pageSize],
  );

  private readonly visibleConfig = this.options.oneWay('pager.visible');

  private readonly visible = computed(
    (visibleConfig, pageCount) => isVisible(visibleConfig, pageCount),
    [this.visibleConfig, this.dataController.pageCount],
  );

  constructor(
    private readonly dataController: DataController,
    private readonly options: OptionsController,
  ) {
    super();
  }

  protected override getProps(): SubsGets<PagerProps> {
    return combined({
      allowedPageSizes: this.allowedPageSizes,
      visible: this.visible,
      pageIndex: computed(
        // TODO: fix the '??'
        (pageIndex) => (pageIndex ?? 0) + 1,
        [this.dataController.pageIndex],
      ),
      pageIndexChanged: (value): void => this.dataController.pageIndex.update(value - 1),
      pageSize: this.dataController.pageSize,
      pageSizeChanged: this.dataController.pageSize.update,
      gridCompatibility: false,
      pageCount: this.dataController.pageCount,
      showPageSizeSelector: this.options.oneWay('pager.showPageSizeSelector'),
      _skipValidation: true,
    });
  }
}
