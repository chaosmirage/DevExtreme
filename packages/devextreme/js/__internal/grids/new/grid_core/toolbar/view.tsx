/* eslint-disable spellcheck/spell-checker */
import type { Subscribable } from '@ts/core/reactive/index';
import { combined, computed } from '@ts/core/reactive/index';

import { View } from '../core/view';
import { OptionsController } from '../options_controller/options_controller';
import { ToolbarController } from './controller';
import { ToolbarView as Toolbar } from './toolbar';
import type { ToolbarProps } from './types';
import { isVisible } from './utils';

export class ToolbarView extends View<ToolbarProps> {
  protected override component = Toolbar;

  private readonly visibleConfig = this.options.oneWay('toolbar.visible');

  private readonly visible = computed(
    (visibleConfig, pageCount) => isVisible(visibleConfig, pageCount),
    [this.visibleConfig, this.controller.items],
  );

  public static dependencies = [ToolbarController, OptionsController] as const;

  constructor(
    private readonly controller: ToolbarController,
    private readonly options: OptionsController,
  ) {
    super();
  }

  protected override getProps(): Subscribable<ToolbarProps> {
    return combined({
      visible: this.visible,
      items: this.controller.items,
      disabled: this.options.oneWay('toolbar.disabled'),
    });
  }
}
