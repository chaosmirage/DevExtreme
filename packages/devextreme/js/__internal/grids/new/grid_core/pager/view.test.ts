/* eslint-disable spellcheck/spell-checker */
/* eslint-disable @typescript-eslint/dot-notation */
import { describe, expect, it } from '@jest/globals';

import { DataController } from '../data_controller/data_controller';
import type { Options } from '../options';
import { OptionsControllerMock } from '../options_controller/options_controller.mock';
import { PagerView } from './view';

const createPagerView = (options?: Options) => {
  const rootElement = document.createElement('div');
  const optionsController = new OptionsControllerMock(options ?? {
    dataSource: [],
    pager: {
      visible: true,
    },
  });

  const dataController = new DataController(optionsController);
  const pager = new PagerView(dataController, optionsController);

  pager.render(rootElement);

  return {
    rootElement,
    optionsController,
  };
};

describe('render', () => {
  it('empty PagerView', () => {
    const { rootElement } = createPagerView();

    expect(rootElement).toMatchSnapshot();
  });

  it('PagerView with options', () => {
    const { rootElement } = createPagerView({
      dataSource: [...new Array(20)].map((_, index) => ({ field: `test_${index}` })),
      paging: {
        pageIndex: 2,
      },
      pager: {
        showPageSizeSelector: true,
      },
    });

    expect(rootElement).toMatchSnapshot();
  });
});

describe('Applying options', () => {
  describe('when visible = \'auto\' and pageCount <= 1', () => {
    it('Pager should be hidden', () => {
      const { rootElement } = createPagerView({
        dataSource: [...new Array(4)].map((_, index) => ({ field: `test_${index}` })),
        paging: {
          pageIndex: 6,
        },
        pager: {
          visible: 'auto',
          showPageSizeSelector: true,
        },
      });

      expect(rootElement).toMatchSnapshot();
    });
  });

  describe('when visible = \'auto\' and pageCount > 1', () => {
    it('Pager should be visible', () => {
      const { rootElement } = createPagerView({
        dataSource: [...new Array(20)].map((_, index) => ({ field: `test_${index}` })),
        paging: {
          pageIndex: 6,
        },
        pager: {
          visible: 'auto',
          showPageSizeSelector: true,
        },
      });

      expect(rootElement).toMatchSnapshot();
    });
  });

  describe('when visible = \'true\'', () => {
    it('Pager should be visible', () => {
      const { rootElement } = createPagerView({
        dataSource: [...new Array(20)].map((_, index) => ({ field: `test_${index}` })),
        paging: {
          pageIndex: 6,
        },
        pager: {
          visible: true,
          showPageSizeSelector: true,
        },
      });

      expect(rootElement).toMatchSnapshot();
    });
  });

  describe('when visible = \'false\'', () => {
    it('Pager should be hidden', () => {
      const { rootElement } = createPagerView({
        dataSource: [...new Array(20)].map((_, index) => ({ field: `test_${index}` })),
        paging: {
          pageIndex: 6,
        },
        pager: {
          visible: false,
          showPageSizeSelector: true,
        },
      });

      expect(rootElement).toMatchSnapshot();
    });
  });

  describe('when changing a visible to \'false\' at runtime', () => {
    it('Pager should be hidden', () => {
      const { rootElement, optionsController } = createPagerView({
        dataSource: [...new Array(4)].map((_, index) => ({ field: `test_${index}` })),
        paging: {
          pageIndex: 6,
        },
        pager: {
          visible: true,
          showPageSizeSelector: true,
        },
      });

      optionsController.option('pager.visible', false);

      expect(rootElement).toMatchSnapshot();
    });
  });

  describe('when changing a visible to \'true\' at runtime', () => {
    it('Pager should be visible', () => {
      const { rootElement, optionsController } = createPagerView({
        dataSource: [...new Array(4)].map((_, index) => ({ field: `test_${index}` })),
        paging: {
          pageIndex: 6,
        },
        pager: {
          visible: false,
          showPageSizeSelector: true,
        },
      });

      optionsController.option('pager.visible', true);

      expect(rootElement).toMatchSnapshot();
    });
  });

  describe('when allowedPageSizes = \'auto\'', () => {
    it('calculates pageSizes by pageSize', () => {
      const { rootElement } = createPagerView({
        dataSource: [...new Array(4)].map((_, index) => ({ field: `test_${index}` })),
        paging: {
          pageIndex: 6,
        },
        pager: {
          visible: true,
          allowedPageSizes: 'auto',
          showPageSizeSelector: true,
        },
      });

      expect(rootElement).toMatchSnapshot();
    });
  });

  describe('when allowedPageSizes with custom values', () => {
    it('displays custom values', () => {
      const { rootElement } = createPagerView({
        dataSource: [...new Array(20)].map((_, index) => ({ field: `test_${index}` })),
        paging: {
          pageIndex: 6,
        },
        pager: {
          visible: 'auto',
          allowedPageSizes: [4, 10, 20],
          showPageSizeSelector: true,
        },
      });

      expect(rootElement).toMatchSnapshot();
    });
  });

  describe('when changing an allowedPageSizes to custom values at runtime', () => {
    it('applies custom values', () => {
      const { rootElement, optionsController } = createPagerView({
        dataSource: [...new Array(20)].map((_, index) => ({ field: `test_${index}` })),
        paging: {
          pageIndex: 6,
        },
        pager: {
          allowedPageSizes: 'auto',
          showPageSizeSelector: true,
        },
      });

      optionsController.option('pager.allowedPageSizes', [4, 10, 20]);

      expect(rootElement).toMatchSnapshot();
    });
  });
});
