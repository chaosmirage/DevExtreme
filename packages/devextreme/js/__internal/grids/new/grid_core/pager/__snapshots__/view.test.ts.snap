// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Applying options when allowedPageSizes = 'auto' calculates pageSizes by pageSize 1`] = `
<div>
  <div
    aria-label="Page navigation"
    class="dx-widget dx-pagination"
    role="navigation"
  >
    <div
      class="dx-page-sizes"
    >
      <div
        aria-label="Items per page: 3"
        class="dx-page-size dx-first-child"
        role="button"
        tabindex="0"
      >
        3
      </div>
      <div
        aria-label="Items per page: 6"
        class="dx-page-size dx-selection"
        role="button"
        tabindex="0"
      >
        6
      </div>
      <div
        aria-label="Items per page: 12"
        class="dx-page-size"
        role="button"
        tabindex="0"
      >
        12
      </div>
    </div>
    <div
      class="dx-pages"
      style="visibility: hidden;"
    >
      <div
        class="dx-page-indexes"
      >
        <div
          aria-label="Page 1"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          1
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Applying options when allowedPageSizes with custom values displays custom values 1`] = `
<div>
  <div
    aria-label="Page navigation"
    class="dx-widget dx-pagination"
    role="navigation"
  >
    <div
      class="dx-page-sizes"
    >
      <div
        aria-label="Items per page: 4"
        class="dx-page-size dx-first-child"
        role="button"
        tabindex="0"
      >
        4
      </div>
      <div
        aria-label="Items per page: 10"
        class="dx-page-size"
        role="button"
        tabindex="0"
      >
        10
      </div>
      <div
        aria-label="Items per page: 20"
        class="dx-page-size"
        role="button"
        tabindex="0"
      >
        20
      </div>
    </div>
    <div
      class="dx-pages"
    >
      <div
        class="dx-page-indexes"
      >
        <div
          aria-label="Page 1"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          1
        </div>
        <div
          aria-label="Page 2"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          2
        </div>
        <div
          aria-label="Page 3"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          3
        </div>
        <div
          aria-label="Page 4"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          4
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Applying options when changing a visible to 'false' at runtime Pager should be hidden 1`] = `
<div>
  
</div>
`;

exports[`Applying options when changing a visible to 'true' at runtime Pager should be visible 1`] = `
<div>
  <div
    aria-label="Page navigation"
    class="dx-widget dx-pagination"
    role="navigation"
  >
    <div
      class="dx-page-sizes"
    >
      <div
        aria-label="Items per page: 3"
        class="dx-page-size dx-first-child"
        role="button"
        tabindex="0"
      >
        3
      </div>
      <div
        aria-label="Items per page: 6"
        class="dx-page-size dx-selection"
        role="button"
        tabindex="0"
      >
        6
      </div>
      <div
        aria-label="Items per page: 12"
        class="dx-page-size"
        role="button"
        tabindex="0"
      >
        12
      </div>
    </div>
    <div
      class="dx-pages"
      style="visibility: hidden;"
    >
      <div
        class="dx-page-indexes"
      >
        <div
          aria-label="Page 1"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          1
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Applying options when changing an allowedPageSizes to custom values at runtime applies custom values 1`] = `
<div>
  <div
    aria-label="Page navigation"
    class="dx-widget dx-pagination"
    role="navigation"
  >
    <div
      class="dx-page-sizes"
    >
      <div
        aria-label="Items per page: 4"
        class="dx-page-size dx-first-child"
        role="button"
        tabindex="0"
      >
        4
      </div>
      <div
        aria-label="Items per page: 10"
        class="dx-page-size"
        role="button"
        tabindex="0"
      >
        10
      </div>
      <div
        aria-label="Items per page: 20"
        class="dx-page-size"
        role="button"
        tabindex="0"
      >
        20
      </div>
    </div>
    <div
      class="dx-pages"
    >
      <div
        class="dx-page-indexes"
      >
        <div
          aria-label="Page 1"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          1
        </div>
        <div
          aria-label="Page 2"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          2
        </div>
        <div
          aria-label="Page 3"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          3
        </div>
        <div
          aria-label="Page 4"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          4
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Applying options when visible = 'auto' and pageCount <= 1 Pager should be hidden 1`] = `
<div>
  
</div>
`;

exports[`Applying options when visible = 'auto' and pageCount > 1 Pager should be visible 1`] = `
<div>
  <div
    aria-label="Page navigation"
    class="dx-widget dx-pagination"
    role="navigation"
  >
    <div
      class="dx-page-sizes"
    >
      <div
        aria-label="Items per page: 3"
        class="dx-page-size dx-first-child"
        role="button"
        tabindex="0"
      >
        3
      </div>
      <div
        aria-label="Items per page: 6"
        class="dx-page-size dx-selection"
        role="button"
        tabindex="0"
      >
        6
      </div>
      <div
        aria-label="Items per page: 12"
        class="dx-page-size"
        role="button"
        tabindex="0"
      >
        12
      </div>
    </div>
    <div
      class="dx-pages"
    >
      <div
        class="dx-page-indexes"
      >
        <div
          aria-label="Page 1"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          1
        </div>
        <div
          aria-label="Page 2"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          2
        </div>
        <div
          aria-label="Page 3"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          3
        </div>
        <div
          aria-label="Page 4"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          4
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Applying options when visible = 'false' Pager should be hidden 1`] = `
<div>
  
</div>
`;

exports[`Applying options when visible = 'true' Pager should be visible 1`] = `
<div>
  <div
    aria-label="Page navigation"
    class="dx-widget dx-pagination"
    role="navigation"
  >
    <div
      class="dx-page-sizes"
    >
      <div
        aria-label="Items per page: 3"
        class="dx-page-size dx-first-child"
        role="button"
        tabindex="0"
      >
        3
      </div>
      <div
        aria-label="Items per page: 6"
        class="dx-page-size dx-selection"
        role="button"
        tabindex="0"
      >
        6
      </div>
      <div
        aria-label="Items per page: 12"
        class="dx-page-size"
        role="button"
        tabindex="0"
      >
        12
      </div>
    </div>
    <div
      class="dx-pages"
    >
      <div
        class="dx-page-indexes"
      >
        <div
          aria-label="Page 1"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          1
        </div>
        <div
          aria-label="Page 2"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          2
        </div>
        <div
          aria-label="Page 3"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          3
        </div>
        <div
          aria-label="Page 4"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          4
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`render PagerView with options 1`] = `
<div>
  <div
    aria-label="Page navigation"
    class="dx-widget dx-pagination"
    role="navigation"
  >
    <div
      class="dx-page-sizes"
    >
      <div
        aria-label="Items per page: 3"
        class="dx-page-size dx-first-child"
        role="button"
        tabindex="0"
      >
        3
      </div>
      <div
        aria-label="Items per page: 6"
        class="dx-page-size dx-selection"
        role="button"
        tabindex="0"
      >
        6
      </div>
      <div
        aria-label="Items per page: 12"
        class="dx-page-size"
        role="button"
        tabindex="0"
      >
        12
      </div>
    </div>
    <div
      class="dx-pages"
    >
      <div
        class="dx-page-indexes"
      >
        <div
          aria-label="Page 1"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          1
        </div>
        <div
          aria-label="Page 2"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          2
        </div>
        <div
          aria-current="page"
          aria-label="Page 3"
          class="dx-page dx-selection"
          role="button"
          tabindex="0"
        >
          3
        </div>
        <div
          aria-label="Page 4"
          class="dx-page"
          role="button"
          tabindex="0"
        >
          4
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`render empty PagerView 1`] = `
<div>
  <div
    aria-label="Page navigation"
    class="dx-widget dx-pagination"
    role="navigation"
  />
</div>
`;
